### PR TITLE
enhancement(enriching): add between date search to enrichment tables

### PR DESCRIPTION
**Pull request by StephenWakely**
ref #8815 

This PR adds two features to enrichment tables.

1. We now have the concept of schemas. Columns are coerced to the suggested type on load. A config can look like this:

```toml
[enrichment_tables.file]
type = "file"
file.path = "../data.csv"
file.encoding.type = "csv"
schema.id = "integer"
schema.dob = "date"
```

Available types are currently -   String,    Date,    DateTime,    Integer,    Float,    Boolean.

Vector doesn't really have the concept of dates without times, so Date is essentially the given DateTime at midnight UTC.

Exact search conditions can now be of any Value type, before it had to be strings.

2. We can now perform a between date search. An example call in VRL is as follows:

```
  data, err = get_enrichment_table_record("file",
    { "id": .lookup,
      "dob": { "from": from,
               "to": to,
	     }
    })
```

Currently this is only supporting dates, but we should follow up with numeric and other comparisons.

It should be noted that performancewise this is currently a very risky maneuver.  In order to perform this search, we index into all fields that have an exact match, then with this subset we perform a sequential scan.

This is ok for the initial user that is planning to use this since they assure us that all of their searches will contain a lot of exact matches in the condition, so there will be at most a handful of results that will need to be scanned through. Benchmarking the scenario where the index search returns 10 records that need scanning through indicates that this takes the whole search from 2μs to 3μs. This is an acceptable tradeoff.

However, there is currently nothing to stop a user from conducting a date search on the entire dataset with no exact match conditions to narrow down the result set. This would result in a sequential scan over the entire dataset. At 1 million rows I was getting benchmark figures of around half a second for this. That wold kill the performance. At minimum, we need documentation to highlight this.

### CODE REVIEW

**Comment by spencergilbert** [comment]
Noting the performance footgun - do we log how long the lookup takes?

File: src/enrichment_tables/file.rs [hunk 1]
```diff
@@ -72,12 +122,24 @@ impl EnrichmentTableConfig for FileConfig {
         } else {
             // If there are no headers in the datafile we make headers as the numerical index of
             // the column.
-            match data.get(0) {
-                Some(row) => (0..row.len()).map(|idx| idx.to_string()).collect(),
-                None => Vec::new(),
+            let mut dataiter = reader.records();
+            match dataiter.next() {
```
Comment (spencergilbert) [inline src/enrichment_tables/file.rs:126]:
```suggestion
            let mut data_iter = reader.records();
            match data_iter.next() {
```

File: benches/enrichment_tables_file.rs [hunk 2]
```diff
@@ -9,65 +11,93 @@ criterion_group!(
 );
 criterion_main!(benches);
 
+/// Returns the text of the column at the given position
```
Comment (pablosichert) [inline benches/enrichment_tables_file.rs:14]:
```suggestion
/// Returns the text of the column at the given position.
```

File: benches/enrichment_tables_file.rs [hunk 3]
```diff
@@ -9,65 +11,93 @@ criterion_group!(
...
+fn column(col: usize, row: usize) -> Value {
+    if col == 0 {
+        // A column that is duplicated across 10 rows
```
Comment (pablosichert) [inline benches/enrichment_tables_file.rs:17]:
```suggestion
        // A column that is duplicated across 10 rows.
```

File: benches/enrichment_tables_file.rs [hunk 4]
```diff
@@ -9,65 +11,93 @@ criterion_group!(
...
+}
+
 fn benchmark_enrichment_tables_file(c: &mut Criterion) {
     let mut group = c.benchmark_group("enrichment_tables_file");
 
-    let setup = |size| {
+    let setup = |size, date_range| {
+        let data = (0..size)
+            .map(|row| {
+                // Add 8 columns
```
Comment (pablosichert) [inline benches/enrichment_tables_file.rs:34]:
```suggestion
                // Add 8 columns.
```

File: lib/vrl/compiler/src/value/convert.rs [hunk 5]
```diff
@@ -230,6 +230,27 @@ impl Value {
             }),
         }
     }
+
+    /// Converts the Value into a byte representation regardless of it's original type.
+    /// Object and Array are currently not supported, although technically there's no reason why it
+    /// couldn't in future should the need arise.
```
Comment (pablosichert) [inline lib/vrl/compiler/src/value/convert.rs:236]:
```suggestion
    /// Converts the Value into a byte representation regardless of its original type.
    /// Object and Array are currently not supported, although technically there's no reason why it
    /// couldn't in future should the need arise.
```

**Review by pablosichert — COMMENTED** [review]
First a small set of nits:

File: src/enrichment_tables/file.rs [hunk 6]
```diff
@@ -114,16 +176,34 @@ impl File {
...
         condition.iter().all(|condition| match condition {
             Condition::Equals { field, value } => match self.column_index(field) {
                 None => false,
-                Some(idx) => row[idx].to_lowercase() == value.to_lowercase(),
+                Some(idx) => match (&row[idx], value) {
+                    (Value::Bytes(bytes1), Value::Bytes(bytes2)) => {
+                        match (std::str::from_utf8(bytes1), std::str::from_utf8(bytes2)) {
+                            (Ok(s1), Ok(s2)) => s1.to_lowercase() == s2.to_lowercase(),
+                            _ => bytes1 == bytes2,
+                        }
```
Comment (pablosichert) [inline src/enrichment_tables/file.rs:191]: I guess you could a case to bail out early:

```suggestion
                        match (std::str::from_utf8(bytes1), std::str::from_utf8(bytes2)) {
                            (Ok(s1), Ok(s2)) => s1.to_lowercase() == s2.to_lowercase(),
                            (Err(_), Err(_)) => bytes1 == bytes2,
                            _ => false,
                        }
```

File: lib/enrichment/src/lib.rs [hunk 7]
```diff
@@ -3,19 +3,26 @@ pub mod tables;
...
 
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug, PartialEq)]
 pub enum Condition<'a> {
-    Equals { field: &'a str, value: String },
+    Equals {
+        field: &'a str,
+        value: Value,
+    },
+    BetweenDates {
```
Comment (pablosichert) [inline lib/enrichment/src/lib.rs:21]: If I judge from the implementation correctly, this condition includes its bounds. Would add a comment here to remove any potential ambiguity

File: src/enrichment_tables/file.rs [hunk 8]
```diff
@@ -169,7 +253,63 @@ impl File {
...
+fn hash_value(hasher: &mut seahash::SeaHasher, value: &Value) -> Result<(), String> {
+    match value {
+        Value::Bytes(bytes) => {
+            hasher.write(
+                std::str::from_utf8(bytes)
+                    .map_err(|_| "column contains invalid utf".to_string())?
+                    .to_lowercase()
+                    .as_bytes(),
+            );
+        }
```
Comment (pablosichert) [inline src/enrichment_tables/file.rs:289]: Looking at the case below, isn't the `hasher` already able to `write` `bytes::Bytes`? Is it necessary to validate for UTF-8 conformance here?

File: src/enrichment_tables/file.rs [hunk 9]
```diff
@@ -169,7 +253,63 @@ impl File {
 
         index.shrink_to_fit();
 
-        index
+        Ok(index)
+    }
+
+    /// Sequentially searches through the iterator for the given condition
```
Comment (pablosichert) [inline src/enrichment_tables/file.rs:259]:
```suggestion
    /// Sequentially searches through the iterator for the given condition.
```

File: src/enrichment_tables/file.rs [hunk 10]
```diff
@@ -114,16 +176,34 @@ impl File {
         self.headers.iter().position(|header| header == col)
     }
 
-    fn row_equals(&self, condition: &[Condition], row: &[String]) -> bool {
+    /// Currently all matches are case insensitive.
+    /// TODO We want to add a configuration option to allow for case sensitive searches. This will
```
Comment (pablosichert) [inline src/enrichment_tables/file.rs:180]: Do we have a follow-up issue / item on the check list on the enrichment table issue for this?

**Comment by pablosichert** [comment]
> It should be noted that performancewise this is currently a very risky maneuver. In order to perform this search, we index into all fields that have an exact match, then with this subset we perform a sequential scan.

If we want to improve the performance for this case, we could use a B+-Tree for range queries here

**Comment by pablosichert** [comment]
I suppose another area for performance improvement would be to change the `File` to hold an array of columns rather than an array of rows, as searching through columns them gives better cache locality, especially for files that have a lot of columns and when only few of them are checked in the condition.

File: lib/vrl/compiler/src/value/convert.rs (hunk 5 above)
Comment (StephenWakely) [reply lib/vrl/compiler/src/value/convert.rs:236]: I believe the apostrophe is correct. We are referring to the original type belonging to it. Thus it is a [possesive apostrophe](https://www.apostrophechecker.com/rules-for-possessive-apostrophes/), and should be included.

File: src/enrichment_tables/file.rs (hunk 8 above)
Comment (StephenWakely) [reply src/enrichment_tables/file.rs:289]: It is needed in order to convert to lowercase that is there. I have another PR coming up that makes the case insensitive search optional, for that case I wouldn't need this.

File: src/enrichment_tables/file.rs (hunk 10 above)
Comment (StephenWakely) [reply src/enrichment_tables/file.rs:180]: #8815 They are all kind of bunched together a bit..

File: src/enrichment_tables/file.rs (hunk 8 above)
Comment (pablosichert) [reply src/enrichment_tables/file.rs:289]: Oh, right! Missed that one

File: lib/vrl/compiler/src/value/convert.rs (hunk 5 above)
Comment (pablosichert) [reply lib/vrl/compiler/src/value/convert.rs:236]: Never thought about it that way, that would be definitely logical.

I think the possessive form with "it" is a special case, though:

> It’s is a contraction of “it is” or “it has.” Its is a possessive determiner we use to say that something belongs to or refers to something.

https://www.grammarly.com/blog/its-vs-its/

> It’s and its are among the most commonly confused words. They are pronounced the same, there’s a very small difference in how they’re written, _and it’s also easy to mistake the contraction in it’s for a possessive_.

Comment (StephenWakely) [reply lib/vrl/compiler/src/value/convert.rs:236]: Well I never!

**Review by spencergilbert — APPROVED** [review]

**Review by pablosichert — APPROVED** [review]
Looking good!

**Comment by binarylogic** [comment]
Same note as other PRs. This needs docs changes.

**Comment by StephenWakely** [comment]
> This needs docs changes.

I will add the docs to a separate PR shortly.

**Pull request merged**